### PR TITLE
[WIP] Add support for SBML species assignment rules + rate rules

### PR DESCRIPTION
**Pull request by jvanhoefer**
I implemented species assignment rules + rate rules for species.

This would close issues #857, #732 and #759.

### CODE REVIEW

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         

Clones added
============
- python/amici/sbml_preprocessing.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    Returns:
+
+    Raises:
+    """
+
+    # loop backwards over rules, so that one can delete rules inside the loop!
+    for i in range(sbml_doc.getModel().num_rules - 1, -1, -1):
+
+        rule = sbml_doc.getModel().getListOfRules()[i]
```
Comment (dweindl) [inline python/amici/sbml_preprocessing.py]:
```suggestion
        rule = sbml_doc.getModel().getRule(i)
```

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         

Clones added
============
- python/amici/sbml_preprocessing.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        if rule.isAssignment() and type(rule_variable) == libsbml.Species:
+
+            p = sbml_doc.getModel().createParameter()
+            p.setId(rule_variable.getId())
+            p.setName(rule_variable.getName())
+            p.setUnits(rule_variable.getUnits())
+            p.setConstant(False)
+
+            print(rule_variable.getId())
```
Comment (dweindl) [inline python/amici/sbml_preprocessing.py]: remove

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import libsbml
+
+
+def flatten_sbml(sbml_doc: libsbml.SBMLDocument):
```
Comment (dweindl) [inline python/amici/sbml_preprocessing.py]: `sbml_doc` is not used anywhere. Why not directly pass the model instead of calling `.getModel()` so often?

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by jvanhoefer** [comment]
@dweindl @FFroehlich Could one of you check the error in the SBML semantic test suite? I do not see, what is going wrong there... (Thank you for helping me out!)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by dweindl** [comment]
> @dweindl @FFroehlich Could one of you check the error in the SBML semantic test suite? I do not see, what is going wrong there... (Thank you for helping me out!)

Done. But can't possibly have worked for you locally. Was missing a symlink for the newly added file.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+            p = sbml.createParameter()
+            p.setId(rule_variable.getId())
+            p.setName(rule_variable.getName())
+            p.setUnits(rule_variable.getUnits())
+            p.setConstant(False)
+
+            # remove modifier, if species is modifier of a reaction
+
+            for reaction in sbml.getListOfReactions():
```
[the remark this answers is not in the document]
Comment (jvanhoefer) [reply python/amici/sbml_preprocessing.py]: Then we would have to remember all species assignments and then search for them as modifiers in the reactions. I can do that...

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    reaction.addModifier(
+                        sbml.getElementBySId(species.getId())
+                    )
+            # remove rule
+            rule.removeFromParentAndDelete()
+
+
+def process_species_assignment_rules(sbml: libsbml.Model):
+    """ Reformulate species assignment rules by replacing the species by
+    a parameter.
```
Comment (dweindl) [inline python/amici/sbml_preprocessing.py]: Wait, I think that won't really work. The concentration of the assigned-to species is supposed to be able to change over time. Parameters don't. Wouldn't we have to replace all occurrences of the assigned-to species by the right hand side of the assignment?

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/AMICI/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ICB-DCM/AMICI/pullRequest?prid=4597862&bid=15433743)

File: python/amici/sbml_preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    Arguments:
+        sbml: Instance of libsbml.Model`.
+
+    Returns:
+
+    Raises:
+    """
+
+    # loop backwards over rules, so that one can delete rules inside the loop!
+    for i in range(sbml.num_rules - 1, -1, -1):
```
Comment (FFroehlich) [inline python/amici/sbml_preprocessing.py]: does `reversed(sbml.getListOfRules())`/`reversed(list(sbml.getListOfRules()))`/`reversed(list(rule for rule in sbml.getListOfRules()))` work here? Seems more pythonic.

Comment (jvanhoefer) [reply python/amici/sbml_preprocessing.py]: good point!

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ICB-DCM/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by jvanhoefer** [comment]
I just noticed, that my strategy for the Species Assignment rules will give an error on all tests:

I tried to generate a parameter with the same name as the species and treat the species assignment rule as a parameter assignment rule after that. Now the flattened SBML has less species than the original one, wich will give an error when comparing the simulated results to the test ground truth... 

How do we want to update the strategy? Or do we therefore in this pull request only focus on rate rules and implement the assignment rules in another way?

**Comment by jvanhoefer** [comment]
I just noticed, that my strategy for the Species Assignment rules will give an error on all tests:

I tried to generate a parameter with the same name as the species and treat the species assignment rule as a parameter assignment rule after that. Now the flattened SBML has less species than the original one, wich will give an error when comparing the simulated results to the test ground truth... 

How do we want to update the strategy? Or do we therefore in this pull request only focus on rate rules and implement the assignment rules in another way?

**Comment by dweindl** [comment]
If RateRules work, it may be reasonable to split things up.

> Now the flattened SBML has less species than the original one, wich will give an error when comparing the simulated results to the test ground truth...

Then that has to be adapted in the verification parts of the SBML test suite code. Sounds feasible?!

**Comment by FFroehlich** [comment]
> If RateRules work, it may be reasonable to split things up.
> 
> > Now the flattened SBML has less species than the original one, wich will give an error when comparing the simulated results to the test ground truth...
> 
> Then that has to be adapted in the verification parts of the SBML test suite code. Sounds feasible?!

Have a look at https://github.com/ICB-DCM/AMICI/blob/a4e077f317d2d3eb6f93fcb1a849ddb4ffa7a175/tests/testSBMLSuite.py#L84

Sounds like having a smarter processing of results might be helpful here. That code is pretty old.  https://github.com/ICB-DCM/AMICI/blob/a4e077f317d2d3eb6f93fcb1a849ddb4ffa7a175/python/amici/pandas.py#L114 should be pretty handy, replace `rdata['x'] with the full dataframe and do the comparison column by column. Will still require conversion to amounts though.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ICB-DCM/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- python/amici/sbml_preprocessing.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ICB-DCM/AMICI/pullRequest?prid=4597862&bid=15433743)

**Comment by dweindl** [comment]
@jvanhoefer : How are things going here?

**Comment by jvanhoefer** [comment]
So, short update:

* I now tried to keep only the code for the _Species Rate_ rules and remove all parts for the _Species Assignment_ rules (as we discussed, the strategy for the species assignment rules will need to be different from what I tried here, due to the mapping of ODE state -> Species...)

* I have now running code, **but** the simulation results are incorrect in the unit tests (e.g. test case 33...). 

* To recap my strategy: reformulate a species rate rule `\dot{s} = f(X, \theta)` by a reaction `R: 0 -> S` with rate `f(X, \theta)`. => _What do you think can go wrong there?_ Ammounts vs. Concentrations?

thank you :)

**Comment by paulstapor** [comment]
> _What do you think can go wrong there?_ Ammounts vs. Concentrations?

That's t least a possible explanation... Might be worth to check that...

**Comment by dweindl** [comment]
>     * To recap my strategy: reformulate a species rate rule `\dot{s} = f(X, \theta)` by a reaction `R: 0 -> S` with rate `f(X, \theta)`. => _What do you think can go wrong there?_ Ammounts vs. Concentrations?

I just ran case 33 with the current state of this branch.
x(0) are as expected. so no problem with amounts / concentrations.

In the generated model, the flux for `reaction1` is `0.5*S1`.
So the RateRule is currently interpreted as `k1 = 0.5`, and not as intended as `d/dt k1 = 0.5`.

**Comment by jvanhoefer** [comment]
Now I fixed some issues, still there are problems with this approach, due to scaling of concentrations with compartment sizes...

But at least every commit fewer tests fail...

**Comment by dweindl** [comment]
@jvanhoefer : Is this still a thing, or obsolete after #1036 ?

**Comment by jvanhoefer** [comment]
> @jvanhoefer : Is this still a thing, or obsolete after #1036 ?

This is obsolete and the branch can be deleted.
